Fix `bump-stage0` build failure, and check-build `bump-stage0` in CI

jieyouxu opened the pull request:
This PR bumps the `toml` dependency of the `bump-stage0` tool to `0.8.23`, which AFAICT is the highest `toml` version that's present in the r-l/r workspace's `Cargo.lock` already (so we don't introduce _another_ `toml 0.x.*` series). I added some byte-buffer-to-string intermediary to workaround `toml 0.8.*` not having the `toml 0.9.*` `toml::from_slice` API.

To catch obvious build failures of the `src/tools/bump-stage0` tool early, before we find out it can't even build when we really need it to work.

Contexts:

- https://github.com/rust-lang/rust/pull/146250#issue-3388327410
- [#t-release > Bump stage0 rustfmt separately ("one-off") @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/241545-t-release/topic/Bump.20stage0.20rustfmt.20separately.20.28.22one-off.22.29/near/537916615)

Fixes rust-lang/rust#146252.

jieyouxu committed:
bootstrap: add check step for `src/tools/bump-stage0`

jieyouxu committed:
ci: check `src/tools/bump-stage0` in `pr-check-2`

To catch obvious build failures of the `src/tools/bump-stage0` early,
before we find out it can't even build when we really need it to work.

Kobzol commented on src/bootstrap/src/core/build_steps/check.rs:
```diff
@@ -839,3 +839,9 @@ tool_check_step!(Linkchecker {
+tool_check_step!(BumpStage0 {
+    path: "src/tools/bump-stage0",
+    mode: |_builder| Mode::ToolBootstrap,
+    default: false
```
It would be nice to have the option to check these tools by default on CI, so that we don't have to enumerate them manually in CI scripts.

Kobzol reviewed (commented):
Thanks!

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146263/397f93362974d298b79e0e0cd43677014aa7b722..7b9980236cc8c32fa1ae5ce6886163a9d7f9ee16)*

Kobzol commented:
@bors r+ rollup

Kobzol commented:
Ah lol, I thought it was fixed already :D Sorry.

jieyouxu committed:
bump-stage0: pick highest common `toml` version, add a workaround

- We pick the higest common `toml` version used in the r-l/r workspace
  to avoid introducing Yet Another `toml` `0.x` version, which happens
  to be `0.8.23` as of the time of writing.
- We introduce a byte-buffer-to-string workaround for the `toml 0.8.*`
  series that do not have the `toml 0.9.*` series's `toml::from_slice`
  API yet. Not efficient, but this is not perf-critical so it's fine.

Kobzol commented:
Looks fine. You can r=me once CI is green.

bors commented:
:pushpin: Commit 312a1240a3cde43ca16e5275457688e47d7bccd8 has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 312a1240a3cde43ca16e5275457688e47d7bccd8 -->
<!-- homu: {"type":"Approved","sha":"312a1240a3cde43ca16e5275457688e47d7bccd8","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
